### PR TITLE
commitlog: Ensure we don't run continuation (task switch) with queues modified

### LINKED ISSUE #9955: Crash in commitlog: Segfault after failed update schema_version: mutation_write_timeout_exception
*Installation details*
Kernel version: `5.11.0-1022-aws`
Scylla version (or git commit hash): `4.6.rc3-0.20220120.f85cd289b with build-id 523535d90f746d46ce5b287fc2f2db8156b5ce19`
Cluster size: 6 nodes (r5b.2xlarge)
Scylla running with shards number (live nodes):
   longevity-100gb-4h-4-6-db-node-dd5b08cf-1 (3.250.33.193 | 10.0.2.11): 8 shards
   longevity-100gb-4h-4-6-db-node-dd5b08cf-2 (34.241.3.99 | 10.0.1.25): 8 shards
   longevity-100gb-4h-4-6-db-node-dd5b08cf-3 (3.250.140.210 | 10.0.3.38): 8 shards
   longevity-100gb-4h-4-6-db-node-dd5b08cf-4 (3.250.193.67 | 10.0.1.211): 8 shards
   longevity-100gb-4h-4-6-db-node-dd5b08cf-5 (52.208.29.36 | 10.0.1.45): 8 shards
   longevity-100gb-4h-4-6-db-node-dd5b08cf-6 (34.244.126.66 | 10.0.3.92): 8 shards
OS (RHEL/CentOS/Ubuntu/AWS AMI): `ami-0760f7d7d8c8fefd6` (aws: eu-west-1)

Test: `longevity-100gb-4h-ebs-gp3-test`
Test name: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):
- [longevity-100gb-4h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/8584390cb25ea782b18acef4a6d77099be58b0bc/test-cases/longevity/longevity-100gb-4h.yaml)
- [ebs-gp3-2v-4tb-16k-iops.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/9894d5c9bddf5edd80da967f290ec6ad02f24c6a/configurations/ebs/ebs-gp3-2v-4tb-16k-iops.yaml)

**Issue description**

====================================
NoCorruptRepair starts
```
< t:2022-01-23 11:05:01,409 f:file_logger.py  l:89   c:sdcm.sct_events.file_logger p:INFO  > 2022-01-23 11:05:01.407: (DisruptionEvent Severity.NORMAL) period_type=begin event_id=baf9d918-bb4a-479e-8dee-f7803ee5b278: nemesis_name=NoCorruptRepair target_node=Node longevity-100gb-4h-4-6-db-node-dd5b08cf-6 [34.244.126.66 | 10.0.3.92] (seed: False)

```
Running multiple C-S commands to prepare test tables `drop_table_during_repair_ks_{i}`
```
2022-01-23 11:09:05.629: (CassandraStressEvent Severity.NORMAL) period_type=begin event_id=755479f5-d0af-475e-913f-08997bbbd5f1: node=Node longevity-100gb-4h-4-6-loader-node-dd5b08cf-2 [3.250.193.229 | 10.0.3.129] (seed: False)
stress_cmd=cassandra-stress write n=400000 cl=QUORUM -port jmx=6868 -mode native cql3 -schema 'replication(factor=6)' -log interval=5
```

Node 4 has a segfault
```
2022-01-23T11:10:45+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !     ERR |  [shard 0] system_keyspace - update compaction history failed: exceptions::mutation_write_timeout_exception (Operation timed out for system.compaction_history - received only 0 responses from 1 
CL=ONE.): ignored
2022-01-23T11:10:45+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |  [shard 0] compaction - [Compact system_schema.columns 1c399db0-7c3d-11ec-bf25-5e02ef09b9c5] Compacting [/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-312-big-Da
ta.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-328-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-304-big-Data.db:level=0:origin=memtable,/var/l
ib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-288-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-296-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns
-24101c25a2ae3af787c1b40ee1aca33f/md-280-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-272-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md
-264-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-320-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/system_schema/columns-24101c25a2ae3af787c1b40ee1aca33f/md-256-big-Data.db:level=0:origin=comp
action]
2022-01-23T11:10:45+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !     ERR |  [shard 0] storage_service - fail to update schema_version for 10.0.3.92: exceptions::mutation_write_timeout_exception (Operation timed out for system.peers - received only 0 responses from 1 CL
=ONE.)
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO | Segmentation fault on shard 3.
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO | Backtrace:
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x3db4458
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x3de5246
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x4433ed1
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x7f04b72bca1f
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x26af36f
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x3dc6e24
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x3dc8217
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x3de5845
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   0x3d9f45a
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   /opt/scylladb/libreloc/libpthread.so.0+0x9298
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO |   /opt/scylladb/libreloc/libc.so.6+0x100352
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 ! WARNING | show_signal_msg: 23 callbacks suppressed
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO | reactor-3[19189]: segfault at 138 ip 00000000026af370 sp 00007f04b35d3530 error 6 in scylla[f3d000+37f8000]
2022-01-23T11:10:47+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 !    INFO | Code: 08 03 00 00 49 89 ae 10 03 00 00 4c 01 e2 49 89 96 18 03 00 00 49 8b 45 08 4c 8b 6c 24 18 48 8b 6c 24 10 48 8b 0d 48 43 10 02 <48> 89 88 38 01 00 00 48 8b 43 28 48 85 c0 0f 84 81 00 00 00 48 83
Jan 23 11:15:17 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 systemd[1]: scylla-server.service: Main process exited, code=killed, status=11/SEGV
Jan 23 11:15:17 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 systemd[1]: scylla-server.service: Failed with result 'signal'.
Jan 23 11:15:17 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 systemd[1]: scylla-server.service: Scheduled restart job, restart counter is at 1.
```

decoded backtrace
```
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:756
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:786
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:798
 (inlined by) seastar::sigsegv_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3712
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3698
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3694
wasmtime_runtime::traphandlers::unix::trap_handler at wasmtime_runtime.9xv9ixev-cgu.15:?
?? ??:0
std::__atomic_base<long>::load(std::memory_order) const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/atomic_base.h:481
 (inlined by) seastar::lowres_clock_impl::steady_now() at ././seastar/include/seastar/core/lowres_clock.hh:61
 (inlined by) seastar::lowres_clock::now() at ././seastar/include/seastar/core/lowres_clock.hh:119
 (inlined by) db::commitlog::segment::reset_sync_time() at ./db/commitlog/commitlog.cc:651
 (inlined by) db::commitlog::segment_manager::new_segment() at ./db/commitlog/commitlog.cc:1639
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2374
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2783
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2952
operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:4166
 (inlined by) void std::__invoke_impl<void, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_98&>(std::__invoke_other, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_98&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
 (inlined by) std::enable_if<is_invocable_r_v<void, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_98&>, void>::type std::__invoke_r<void, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_98&>(seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_98&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:111
 (inlined by) std::_Function_handler<void (), seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_98>::_M_invoke(std::_Any_data const&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:291
std::function<void ()>::operator()() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:560
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:60
?? ??:0
?? ??:0

```

before the segfault happened we had a lot of `WriteTimeoutException`

```
< t:2022-01-23 11:04:55,371 f:base.py         l:222  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during SIMPLE write query at consistency QUORUM (2 replica were required but only 1 acknowledged the write)
< t:2022-01-23 11:04:55,371 f:base.py         l:222  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during SIMPLE write query at consistency QUORUM (2 replica were required but only 1 acknowledged the write)
< t:2022-01-23 11:04:55,371 f:base.py         l:222  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during SIMPLE write query at consistency QUORUM (2 replica were required but only 1 acknowledged the write)
< t:2022-01-23 11:04:55,371 f:base.py         l:222  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.WriteFailureException: Cassandra failure during write query at consistency QUORUM (2 responses were required but only 1 replica responded, 2 failed)
< t:2022-01-23 11:04:55,371 f:base.py         l:222  c:RemoteCmdRunner      p:DEBUG > com.datastax.driver.core.exceptions.WriteFailureException: Cassandra failure during write query at consistency QUORUM (2 responses were required but only 1 replica responded, 2 failed)

```
and  some `mutation_write_timeout_exception`
```
(DatabaseLogEvent Severity.WARNING) period_type=one-time event_id=8d72cd85-6a74-49f9-bab7-3a31e4e2f591: type=SYSTEM_PAXOS_TIMEOUT regex=(mutation_write_|Operation timed out for system.paxos|Operation failed for system.paxos) line_number=8278 node=longevity-100gb-4h-4-6-db-node-dd5b08cf-4
2022-01-23T11:05:24+00:00 longevity-100gb-4h-4-6-db-node-dd5b08cf-4 ! ERR | [shard 0] storage_service - fail to update schema_version for 10.0.3.38: exceptions::mutation_write_timeout_exception (Operation timed out for system.peers - received only 0 responses from 1 CL=ONE.)
```
Node 4 also experienced failed  and timed out writes before the segfault
image

Also some writes were blocked on commitlog

image

====================================

Restore Monitor Stack command: `$ hydra investigate show-monitor dd5b08cf-5731-4574-ba3a-6e796f1d4e97`
Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=dd5b08cf-5731-4574-ba3a-6e796f1d4e97)
Show all stored logs command: `$ hydra investigate show-logs dd5b08cf-5731-4574-ba3a-6e796f1d4e97`

Test id: `dd5b08cf-5731-4574-ba3a-6e796f1d4e97`

Logs:
grafana - [https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_114518/grafana-screenshot-longevity-100gb-4h-ebs-gp3-test-scylla-per-server-metrics-nemesis-20220123_114700-longevity-100gb-4h-4-6-monitor-node-dd5b08cf-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_114518/grafana-screenshot-longevity-100gb-4h-ebs-gp3-test-scylla-per-server-metrics-nemesis-20220123_114700-longevity-100gb-4h-4-6-monitor-node-dd5b08cf-1.png)
grafana - [https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_114518/grafana-screenshot-overview-20220123_114518-longevity-100gb-4h-4-6-monitor-node-dd5b08cf-1.png](https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_114518/grafana-screenshot-overview-20220123_114518-longevity-100gb-4h-4-6-monitor-node-dd5b08cf-1.png)
db-cluster - [https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/db-cluster-dd5b08cf.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/db-cluster-dd5b08cf.tar.gz)
loader-set - [https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/loader-set-dd5b08cf.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/loader-set-dd5b08cf.tar.gz)
monitor-set - [https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/monitor-set-dd5b08cf.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/monitor-set-dd5b08cf.tar.gz)
sct - [https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/sct-runner-dd5b08cf.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/dd5b08cf-5731-4574-ba3a-6e796f1d4e97/20220123_115712/sct-runner-dd5b08cf.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-4.6/job/longevity/job/longevity-100gb-4h-ebs-gp3-test/10/)

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #9955

In #9348 we handled the problem of failing to delete segment files on disk, and
the need to recompute disk footprint to keep data flow consistent across intermittent
failures. However, because _reserve_segments and _recycled_segments are queues, we
have to empty them to inspect the contents. One would think it is ok for these
queues to be empty for a while, whilst we do some recaclulating, including
disk listing -> continuation switching. But then one (i.e. I) misses the fact
that these queues use the pop_eventually mechanism, which does _not_ handle
a scenario where we push something into an empty queue, thus triggering the
future that resumes a waiting task, but then pop the element immediately, before
the waiting task is run. In fact, _iff_ one does this, not only will things break,
they will in fact start creating undefined behaviour, because the underlying
std::queue<T, circular_buffer> will _not_ do any bounds checks on the pop/push
operations -> we will pop an empty queue, immediately making it non-empty, but
using undefined memory (with luck null/zeroes).

Strictly speakging, seastar::queue::pop_eventually should be fixed to handle
the scenario, but nontheless we can fix the usage here as well, by simply copy
objects and do the calculation "in background" while we potentially start
popping queue again.

### CODE REVIEW

**Comment by bhalevy** [comment]
Cc @nyh

**Comment by bhalevy** [comment]
@elcallio the queue class wasn't design to work this way.
It is meant to be used with a single producer and single consumer:
https://github.com/scylladb/seastar/blob/97fe0383a5778b1fa1fab39db2dcb85672cb9b99/include/seastar/core/queue.hh#L31-L33

**Comment by bhalevy** [comment]
I'm nervous about the direction that you're suggesting since I'm concerned there could be another timing dependent scenario leading to calling `pop()` on an empty queue.
Since this usage is not on the hot path, can we use a semaphore to serialize the consumers?

**Comment by bhalevy** [comment]
Alternatively, you can implement your own pop_eventually that will verify that the queue is not empty after `not_empty()` resumes.

**Comment by elcallio** [comment]
@bhalevy - the code is safe, as long as we maintain the pre-post conditions every time we cross a task switch boundary. 
Imho, queue's api does not really (beyond above comment) discourage usage like:
```
task1:
    queue.push();
    queue.pop();
    <continuation>

task2:
    queue.pop_eventually()-> [eventual break]
``´
It is a trivial fix to queue to make it recurse on the waiting paths and thus handle the scenarios. Adding bounds checks would not exactly be performance detrimental either. 

Of course, being able to const-iterate the contents of the queue would also fix this case. But alas, here we are. 
I don't want to go through seastar changes just to fix a broken usage here. To repeat, the problem with the old code was that we (again to save a few copy-cycles) allowed task switching between pop/push pairs. Removing this fixes the issue.

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -1963,6 +1963,17 @@ future<> db::commitlog::segment_manager::recalculate_footprint() {
...
+        // interruptable/continuation. Because both queues are
+        // used with push/pop eventually which _waits_ for signal
+        // but does _not_ verify that the condition is true once
+        // we return. So copy the objects and look at instead.
+        for (auto& filename : recycles) {
+            _recycled_segments.push(sstring(filename));
+        }
+        for (auto& s : reserves) {
+            _reserve_segments.push(sseg_ptr(s)); // you can have it back now.
+        }
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:1976]: Looks like a preexisting problem, but still. queue::push() return false if the push failed, and here we aren't checking the return value. What happens if we fail to push?

Comment (avikivity) [reply db/commitlog/commitlog.cc:1976]: Ah, I see you just popped them off.

But I think you shouldn't be using seastar::queue here. This isn't decoupling some unrelated pieces of software, they are tightly coupled, and here we're losing by not seeing items waiting in push_eventually(). All we need is plain containers and maybe condition variables for wakeups.

Comment (elcallio) [reply db/commitlog/commitlog.cc:1976]: I agree, but it is a bigger change. This is a minimal fix. 
Also see seastar patch for queue, which I think could benefit from being somewhat more resilient to this (mis)usage pattern.

**Comment by avikivity** [comment]
Backported to 4.6 (only vulnerable release)

**Comment by avikivity** [comment]
@yaronkaikov looks like the bot forgot to label it as a backport candidate, why? Please make sure other commits weren't missed.

**Pull request closed without merging**